akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala

```diff
@@ -102,10 +102,8 @@ private[cluster] class ClusterRemoteWatcher(
    * by super RemoteWatcher.
    */
   def takeOverResponsibility(address: Address): Unit = {
-    watching foreach {
-      case (watchee, watcher) ⇒ if (watchee.path.address == address)
-        unwatchRemote(watchee, watcher)
-    }
+    watching.keys.filter(_.path.address == address).foreach(clearAllWatches)
```

Comment:
keys is lightweight, it creates a wrapper around the map. You're right for filter however.
I propose a `for (watchee <- watching.keys`) with a guard condition (efficient and I find this a bit  more readable at the same time)